GNRC implementation of CoAP

kb2ma opened the pull request:
See [README.md](https://github.com/kb2ma/RIOT/blob/gcoap/master/examples/gcoap/README.md) in examples/gcoap for an overview. Includes a CLI coap command.

Why a GNRC implementation when libcoap and microcoap are out there? Well, it started for personal reasons, as an experiment with the network stack, and I have kept on hacking at it. Certainly CoAP is a key IoT protocol. IMO it's worth the effort to have a well-documented LGPL implementation based on RIOT's messaging architecture, that can be tuned to the project's needs.

To make the implementation more useful, I plan to implement Observe soon. If others agree the the goal is worthwhile and the work looks promising, I'll add unit tests for what I have so far.

BytesGalore commented:
@kb2ma nice one, I remember a discussion (but no details) with @smlng some Hack'n'ACKs ago about the shortcomings of our used *CoAP packages. 
@smlng could be interested in this one.

gebart commented:
I certainly think it is worthwhile if we get a RIOT coap implementation which uses the riot IPC to make it easier to get started on IoT projects. 
What is the current state of this implementation? Does it provide both server and client? 
Block1, block2?

kb2ma commented:
Thanks for the feedback. The implementation includes client and server, but I have focused mostly on client side so far. No block support yet. RIOT's generic IP fragmentation makes block less pressing, although I have not compared the per-packet payload size between those two.

Here is a summary of what's available:
- Server listens on a port, and provides a callback for generation of an application-specific response.
- Client uses gnrc_coap_send() for a request, and listens for a response via a callback. Matches response on IP port and CoAP token.
- Message Type: Supports non-confirmable (NON) messaging.
- Options: Supports URI-Path (ASCII, no percent-encoding) and Content-Format for payload.

gebart commented:
In my opinion there are two issues with relying on the IP fragmentation for block transfers: the size is limited to the maximum size of a IP packet, <1280 bytes, and if one fragment is lost, the whole packet needs to be retransmitted, unless other methods such as the link-layer retransmissions manage to catch the missing fragment frame.
Using a block transfer is usually more costly in terms of bytes total for the message, but more reliable.

kb2ma commented:
I have posted the [slides](http://cytheric.net/files/gcoap.pdf) and accompanying [audio](http://cytheric.net/files/gcoap.mp3) overview of gcoap that I presented Saturday at the Web and IoT discussion for the RIOT Summit.

OlegHahm commented:
@kb2ma, thanks a lot!

gebart commented:
@kb2ma thank you for posting it! I wish I had been able to attend the summit. I will listen to this later tonight!

OlegHahm commented:
@kb2ma, can you point me to the function/data structure one needs to implement in order to generate a resource and give it a name?

kb2ma commented:
@OlegHahm, good to hear from you! It sounds like you want to implement a server, and register a resource to some path. I have not implemented this registration, but gcoap does have a proto-solution.

All of the structs for gcoap are defined in coap.h, and there is some introductory documentation at the top of that file. See especially the Server Use section. The 'server struct' referenced there is gnrc_coap_server_t. 

So, the proto-solution is a callback that allows a user of gcoap to implement their own resource server. A super-simple server could be implemented in a single function. See the gcoap example for a server that only provides an empty response to /.well-known/core. Now that's simple. :-) The actual resource transferred is the gnrc_coap_transfer_t.

At this point, anything more sophisticated is up to you. To date, I personally have been more focused on the client side of CoAP. However, I am starting on the goal of implementing the Observe protocol. For a working example, I plan to start with your netstats work (yes!) as a resource to serve, as I am interested in network monitoring.

This area of resources is tricky, and really server-specific. Certainly we don't want unnecessary storage at the CoAP layer. At a high level, what are your expectations from a server for resource registration -- pass a pointer, a callback, store some data?

Monday is RIOT day for me this week, so let me try to come up with some concrete ideas for serving up the netstats data.

OlegHahm commented on sys/net/gnrc/application_layer/coap/gnrc_coap.c:
```diff
@@ -0,0 +1,542 @@
+    coap_type          = (hdr->ver_type_tkl & 0x30) >> 4;
+    msg_meta->tokenlen =  hdr->ver_type_tkl & 0x0F;
+    if (coap_ver != GNRC_COAP_VERSION 
+            || coap_type != GNRC_COAP_TYPE_NON
```
So, only non-confirmable messages are supported right now?

smlng commented:
@OlegHahm: yes, only non-con - as Ken mentioned above:

> Message Type: Supports non-confirmable (NON) messaging.

and during his presentation at the summit.

OlegHahm commented:
Oops, sorry, too little sleep during the last days. ;)

kb2ma commented:
No worries, @OlegHahm, and thanks for stepping in, @smlng. Yes, this PR definitely is WIP. I have added [timeouts](https://github.com/kb2ma/RIOT/commit/93c7ed6a67c32e7974d115c2fc4563751c6f8a91) in my working branch, which are essential for confirmable messaging. BTW, is there a best practice for refreshing the PR? I was planning to wait until I have some significant feature ready.

My thinking on confirmable has been that any failures in the mesh mean that some link layer retries already failed. Besides, if the message is really important, an app needs a fallback strategy anyway, in case confirmable retries fail, too. I definitely agree though that the ACK confirmation is nice to have.

gebart commented:
@kb2ma while you have the WIP tag on the PR you should be able to push whatever you want to whenever you want to. It's always nice to see things moving ahead. You should probably rebase regularily while working on it to stay up to date with the master branch.
When you are ready for a review you should let us know in the comments and edit the title to remove the "WIP". After that you should avoid squashing without first asking the reviewers that it is OK.

kaspar030 commented:
I must say that I'm not sure the tight coupling of CoAP protocol handling and gnrc is good design. A proper CoAP library should sit below any gnrc layer.

(I'm also working on a CoAP library here [1], functionality wise it's not further than this PR.)

[1] https://github.com/kaspar030/sock/tree/master/nanocoap

kb2ma commented:
Thanks for the feedback and link to nanocoap, @kaspar030. Have you defined features and timeline for it? Are you planning to add Observe, Confirmable, or Block in the near term?

gebart commented:
@kaspar030 @kb2ma you should join forces, and try to unify your implementations.

I think using `conn` for the network layer interactions makes sense, CoAP is a session layer protocol (OSI) and it makes sense to try to use the abstractions provided by `conn` or `sock` (hopefully, `conn` and `sock` will be the same thing after #5533 is resolved)

OlegHahm commented:
While I agree with @kaspar030 and @gebart on the one hand (putting CoAP on top of conn/sock/sockets increases flexibility a lot), I see problems with that:
1. You would have either to wait until our discussion in #5533 has concluded (and maybe provide input) or use the current conn interface and update it later.
2. GNRC has IMO a very nice interface and sticking to it may lead to a better design in the end.

Moreover, CoAP cannot only run over UDP, but a plethora of protocols (the two craziest things I heard about were CoAP-over-SMS and CoAP-over-WebRTC). Hence, netapi might be a better match to enable this flexibility than conn.

kb2ma commented:
I agree that refitting this PR to work on conn/sock is hung up on #5533. Having said that, I expect refitting will be easier if conn/sock adds an async event loop, as is being discussed now. For my implementation, this feature (plus message passing) is the main appeal of GNRC. I can live with the ambiguity for now, and will see how the other PR resolves.

kaspar030 commented:
> Thanks for the feedback and link to nanocoap, @kaspar030. Have you defined features and timeline for it? Are you planning to add Observe, Confirmable, or Block in the near term?

Those features are planned, but I don't know when I find time.
The rough roadmap is:
- get conn/sock to a usable state
- replace microcoap with nanocoap
- then gradually add the missing features

I don't see nanocoap tied to conn/sock though. It's parsing and state handling will be independent from the used networking. I think it would make sense to at least share that code, if possible.

kb2ma commented:
Sounds like a good plan. I reviewed the nanocoap code, and I think it works well as a microcoap replacement. It definitely is the thinnest veneer over the packet data.

I agree that the packet handling code should be isolated from the networking code. For example, my _coap_parse() expects a gnrc_pktsnip_t, but it just needs a buffer and length as you have done.

It looks like nanocoap really is a few things: 

1. a network application encoded in main.c
2. a server library encoded in the endpoint struct/array and coap_build_reply(), as used in handler.c
3. a packet library in coap_parse(), and the static getters and setters, as well as the hdr and pkt structs

Let me think on this list from the bottom up, for how these things can be integrated with gcoap.

kb2ma commented:
I see a way forward now with use of nanocoap for gcoap. I plan to make these migrations for the structs in gcoap. The structs don't match one-to-one, but this is the basic idea.

| nancoap         | gcoap/description |
| --------------- | -------------------- |
| coap_hdr_t      | migrate from gnrc_coap_meta_t |
| coap_pkt_t      | migrate from gnrc_coap_transfer_t |
| coap_endpoint_t | add use; provides a way to specify local resources or to access a remote server |
| coap_sender_t   | retain to track sent requests; add a pointer for the target endpoint; add a pointer for the buffer used by coap_hdr_t, probably a pktsnip_t |

At least for development purposes, I plan to make nanocoap a module at the 'net' level, in `sys/net/application_layer/coap`. It will be a dependency for gcoap. I'll make any changes for gcoap to that module, and then Kaspar and I can work out the differences with the nanocoap in his sock repository. I'll bring up specific issues as they arise.

This refactoring is substantial, and I plan to take some vacation time soon, so expect a few weeks for some results.

kb2ma commented:
Making steady progress now on implementation of server operation. I added a wiki and issues to my fork to keep track of the work. See the [Status page](https://github.com/kb2ma/RIOT/wiki/gcoap-Status) for details. I plan to refresh the PR when the server is useful.

kb2ma commented:
I have a working CoAP server now for a reimplementation of gcoap based on Kaspar's nanocoap. As a reimplementation, I refreshed my feature branch with a force push.

To get started, see the [README](https://github.com/kb2ma/RIOT/tree/gcoap/master/examples/gcoap). Be sure to look at the source doc referenced in the README (from [coap.h](https://github.com/kb2ma/RIOT/blob/gcoap/master/sys/include/net/gnrc/coap.h)) for how to write an endpoint callback. 

gcoap still works on GNRC, but I see is a clear path to use of sock if/when the time is right. A user of gcoap will not need to change their endpoint callbacks.

gcoap uses the core of nanocoap: coap_pkt_t, coap_hdr_t, coap_handler_t, and coap_endpoint_t structs, as well as coap_parse() and the packet helper functions. 

gcoap does not use coap_handle_req(), but its _receive() function is structured similarly. gcoap wraps the definition of endpoints into a gcoap_listener_t linked list struct. Each application defines its own listener and endpoints, and then registers the listener with gcoap. See the gcoap CLI example (gcoap_cli.c) at the README link above.

gcoap also does not ask the callback to use coap_build_reply(). Instead it defines gcoap_resp_header() and gcoap_resp_content() for this purpose. The latter is required only when the response includes a payload.

The most time consuming part of the reimplementation was figuring out a simple but flexible way to write the response. CoAP Option tags are sandwiched between the header and the payload and must be in a specific order. This make it difficult to write the packet in order from start to finish within the context of an application callback.

@kaspar030, I made some specific changes to the nanocoap .h and .c. These changes are described on a [wiki page](https://github.com/kb2ma/RIOT/wiki/nanocoap-Notes). Happy to discuss if/how we should incorporate these.

Thanks for your review.

smlng commented:
@kb2ma just a quick note. From what I found nanocoap is a fork of microcoap bringing some optimization. However, it retains (IMO wrong) semantics introduced by microcoap, i.e. `coap_endpoint_t` should be `coap_resource_t`. Its just a minor flaw I know, but an _endpoint_ in CoAP is the one requesting or holding a resource, not the resource it self (see [here](https://tools.ietf.org/html/rfc7252#section-1.2)).

kb2ma commented:
Thanks for the pointer to semantics. I agree that this entity is commonly referred to as a resource rather than an endpoint. My goal has been to adapt to nanocoap based on the comments from July, and it didn't seem like the difference was that important.

I looked at microcoap several months ago, but since it did not include client side capabilities, I did not look that closely. I'll compare with nanocoap and reply back.

kaspar030 commented:
> From what I found nanocoap is a fork of microcoap bringing some
> optimization.

Not a fork, I started nanocoap from scratch.

smlng commented:
@kaspar030 some naming in nanocoap sounded familiar, so I assumed it was a some fork of microcoap. Sorry, my bad - no offense! 

But still, usage of `endpoint` is IMO wrong (or misleading) and should be `resource` instead.

kaspar030 commented:
> But still, usage of |endpoint| is IMO wrong (or misleading) and should
> be |resource| instead.

Yeah, the naming comes from microcoap. Will improve.

kb2ma commented on sys/net/gnrc/application_layer/coap/gcoap.c:
```diff
@@ -0,0 +1,363 @@
+    gcoap_resp_header(&coap_pkt, &buf[0], GCOAP_RESPONSE_BUF_SIZE,
+                                          COAP_CODE_PATH_NOT_FOUND);
+
+send:
```
### GNRC vs. sock
This send: label is where the response packet is finalized and sent to the transport layer. Here you can see how GNRC is actually used, and how we can support sock in the future.

1. _finalize_response() finalizes the buffer after the callback. 
2. Then we allocate a GNRC pktsnip and copy from the buffer into the snip. We use two memcpy() calls because the packet bytes from the callback include space between the header and the payload so _finalize_response() can write the options in order.
3. Finally _send() creates the rest of the packet buffer snips and dispatches.

To support sock, we need to write the final packet contents in the buffer rather than in the pktsnip. In this case, we can use memmove() on the buffer. Then the buffer can be passed to sock as a whole.

kb2ma commented:
I have a working CoAP client now for the reimplementation of gcoap to use Kaspar's nanocoap. Rebased and pushed to my gcoap/master branch. To get started, see the [README](https://github.com/kb2ma/RIOT/tree/gcoap/master/examples/gcoap). Be sure to look at the source doc referenced in the README (from [coap.h](https://github.com/kb2ma/RIOT/blob/gcoap/master/sys/include/net/gnrc/coap.h)) for how to write a request. 

In comparison to nanocoap, gcoap uses asynchronous GNRC messaging to send a client request and wait for a response. The user provides a callback to handle the response or the timeout. gcoap's API also is somewhat higher level. The user is not required to specify the actual options. gcoap writes the options after the user has written the payload.

@kaspar030, thanks for the coap option functions! I made more changes to the nanocoap .h and .c. These changes are described on a [wiki page](https://github.com/kb2ma/RIOT/wiki/nanocoap-Notes). Happy to discuss if/how we should incorporate these.

kb2ma commented:
I just added a commit to improve the alignment of gcoap's 'send' functions with those used by nanocoap/yacoap. At this point, yacoap could use gcoap's gcoap_req_send() for a request, but it would need to provide a callback that accepts a coap_pkt_t. I also added a [wiki page](https://github.com/kb2ma/RIOT/wiki/nanocoap%2C-yacoap%2C-and-gcoap-compared) that compares the functions used by the three libraries.

I have a little more work to do, for example ensuring gcoap's function signatures are consistent in how they reports errors. Then I plan to remove the WIP in the PR title.

OlegHahm commented on sys/net/gnrc/application_layer/coap/gcoap.c:
```diff
@@ -0,0 +1,569 @@
+void gcoap_op_state(uint8_t *open_reqs)
+{
+    uint8_t count = 0;
+    for (uint8_t i = 0; i < GCOAP_REQ_WAITING_MAX; i++) {
```
Please use `int` or `unsigned` for iterator variables (cf. http://stackoverflow.com/questions/34891181/for-array-loop-use-int-or-uint8-t).

We should probably add this to the coding conventions.

kb2ma replied:
Thanks, Oleg. I'll add this to my lint pass through the code.

emmanuelsearch commented:
@kb2ma is it doable in your mind to aim for polishing and merging the current state of gcoap for the next release?

emmanuelsearch commented:
Actually I had another question: have you considered adding basic support for blockwise transfer [(RFC7959)?](https://tools.ietf.org/html/rfc7959) ? This could be useful for firmware image transfer, for example...

kaspar030 commented:
You might be interested in some nanocoap changes I did the other day.

kb2ma commented:
@emmanuelsearch , is there a cutoff date for getting the PR into the release? Yes, I plan to add observe, block, and formal confirmable support. I try to keep this sort of information updated on a  [status page](https://github.com/kb2ma/RIOT/wiki/gcoap-Status) on my wiki clone.

kb2ma commented:
Thanks for the heads-up, @kaspar030. I am working to incorporate the latest changes.

I have a few questions about how to integrate nanocoap for the purpose of merging gcoap. Presently, I have put only nanocoap.c and .h in the sys tree.

1. Is this overall approach acceptable, or does it make more sense to import nanocoap into the pkg tree?
2. If the overall approach is OK, is it acceptable to just include what I need from nanocoap? For example, I have not included nanocoap_sock.c
3. Are you OK with me including these files as part of merging gcoap, or did you want to do this yourself separately? What about unit tests -- are you planning to create them?

kaspar030 commented:
> Presently, I have put only nanocoap.c and .h in the sys tree, and created a module for it. The gcoap module depends on nanocoap.

>     Is this overall approach acceptable, or does it make more sense to import nanocoap into the pkg tree?

I plan on adding nanocoap as pkg. Actually I was waiting for gnrc sock to be merged...
I'll split up nanocoap into modules so one just compiles nanocoap.c to be used by gcoap.

>     If the overall approach is OK, is it acceptable to just include what I need from nanocoap? For example, I have not included nanocoap_sock.c

That should be fine, and I put the sock specific stuff into it's own file for exactly that reason.

>     Are you OK with me including these files as part of merging gcoap, or did you want to do this yourself separately? What about unit tests -- are you planning to create them?

I'd be OK from a personal perspective. Let me look over your nanocoap changes, if we can agree on a direction there, I'd prefer gcoap to use the pkg I intend to create.

Feel free to write unittests! ;)

kb2ma commented:
Use of your pkg for just nanocoap.c sounds best to me. I will plan to use that when it is ready.

Let me incorporate your latest nanocoap work into my copy, and I'll let you know when it's ready for review. I'll work on that this weekend. I have found a couple of bugs/issues, but I have made some stylistic changes that don't necessarily need to remain. I'm sure we can make this work, but if there are any gaps, I can just add them in gcoap.

kb2ma commented:
I see there are 2 1/2 weeks until the freeze. I think that within a week, I can resolve integration with nanocoap and polish the code enough to remove the WIP label. Then the biggest gap I see is that there are no unit tests. So @emmanuelsearch, assuming I can bridge that gap, I will do my best to get this work merged with the release.

kb2ma commented:
@kaspar030, I have updated nanocoap in gcoap based on your latest upstream work. Next, I plan to post issues and PRs against your sock repository to resolve the remaining differences. I have been maintaining a [wiki page](https://github.com/kb2ma/RIOT/wiki/nanocoap-Notes) on the differences, but I think it will be simpler to address them one by one. I plan to start on that later today.

kb2ma commented:
Created a [PR](https://github.com/kaspar030/sock/pull/1) and an [Issue](https://github.com/kaspar030/sock/issues/2) against Kaspar's sock repository. These two items cover all of the changes I have made to nanocoap.h and .c.

kb2ma commented:
With these last two commits, I am happy with the overall structure and functionality of gcoap. From this perspective, I am ready to drop the WIP label. As a matter of policy, I expect the project first would want to see the PR for nanocoap. So, I will wait to change the title until that is resolved.

kaspar030 commented:
> the project first would want to see the PR for nanocoap.

It's PRed now in #5972.

kb2ma commented:
This PR now depends on #5972. Rebased on the branch for that PR. Updated the title for this PR to remove WIP. @kaspar030, please remove the label when ready, and thanks for a smooth transition!

kb2ma commented:
Is there a recipe or published limits so I can populate the BOARD_INSUFFICIENT_MEMORY variable for gcoap's example make file?

kb2ma commented:
I have finished with my refinements to gcoap. The only open question is how to populate BOARD_INSUFFICIENT_MEMORY. I have no plans to add more commits until a review.

kb2ma commented:
Rebased on latest nanocoap pkg and squashed to 3 commits. Actually, it was the latest nanocoap pkg until about a minute before I pushed to GitHub. At any rate, I also have blacklisted nrf52dk, which was the commit from a minute ago.

kaspar030 commented:
> Rebased on latest nanocoap pkg and squashed to 3 commits. Actually, it was the latest nanocoap pkg until about a minute before I pushed to GitHub. At any rate, I also have blacklisted nrf52dk, which was the commit from a minute ago.

Wow, you're fast. :)

I'll be gone for the next ten days, so there won't be any changes. BTW, I've started working on blockwise transfers, maybe we could discuss how to integrate that API-wise when I'm back?

kb2ma commented:
> BTW, I've started working on blockwise transfers, maybe we could discuss how to integrate that API-wise when I'm back?

Sounds good, I'd be happy to.

kb2ma commented:
Now that nanocoap is merged, I will rebase on RIOT master. Maintainers -- the WIP and Needs Squashing labels no longer are valid.

miri64 commented:
\o/ Maybe we can merge it for this release?

kb2ma commented:
That's my goal! Working on it now.

miri64 commented:
(feature freeze is today/tomorrow-ish ;-))

miri64 commented:
(I'm willing to wait a little for this PR (and others) but won't overstretch the timetable)

miri64 commented on examples/gcoap/Makefile:
```diff
@@ -0,0 +1,63 @@
+USEMODULE += gnrc_netdev_default
+USEMODULE += auto_init_gnrc_netif
+# Specify the mandatory networking modules
+USEMODULE += gnrc_ipv6_router_default
```
Maybe you can remove the router requirements on this, too? RPL you already commented out below anyways ;-) (see #6007).

kb2ma replied:
Done on top of rebase.

kb2ma commented:
@miri64, I just read your 'One more day' message. To be clear, I addressed the issue you flagged about configuring the example as a host. I'm not sure what is next, but I expect that you will ask me to squash that. I will do that, and then either we are done or there will be more review. Let me know if there is anything I should do to move the process along.

miri64 commented on tests/unittests/tests-gcoap/Makefile.include:
```diff
@@ -0,0 +1,8 @@
+
+USEMODULE += gcoap
+USEMODULE += gnrc_udp
+USEMODULE += gnrc_ipv6
```
Why are these modules included?

kb2ma replied:
1. gcoap is necessary because we are testing that module. :-)
2. gnrc_ipv6 is needed at least due to code in gcoap's _event_loop(), which references GNRC_NETTYPE_IPV6. This dependency is my assumption. Is there a better alternative?
3. gnrc_udp is not needed, and I will fix that. It is an implicit (Makefile.dep) dependency for gcoap. Thanks for that.

miri64 replied:
I think there are other tests already that include `gnrc_ipv6` so I'm fine with that (`gcoap` was clear ^^")

miri64 replied:
Could you maybe make this more clear (e.g. by adding "uncomment if needed" like in the [`gnrc_minimal`](https://github.com/RIOT-OS/RIOT/blob/master/examples/gnrc_minimal/Makefile#L12) example)?

miri64 commented on examples/gcoap/Makefile.slip:
```diff
@@ -0,0 +1,84 @@
+
+# Redefine request token length, max 8.
+#GCOAP_TOKENLEN = 2
+#CFLAGS += -DGCOAP_TOKENLEN=$(GCOAP_TOKENLEN)
```
Can those be removed or do you plan to uncomment them?

miri64 replied:
https://github.com/RIOT-OS/RIOT/blob/master/examples/gnrc_minimal/Makefile#L12

miri64 replied:
Could you maybe add some comment about than (like "uncomment if needed" like in the [`gnrc_minimal`](https://github.com/RIOT-OS/RIOT/blob/master/examples/gnrc_minimal/Makefile#L12) example)?

miri64 commented:
Just let me give this a quick review before squashing. Also it might be that you will require #6030 to pass Murdock :-/

kb2ma commented:
Thanks. Regarding #6030, the Makefile explicitly imports the posix module due to #5959, and I expect this gives the build what it needs.

miri64 commented:
Yes :-).

kb2ma commented:
Squashed, with gusto!

miri64 commented on tests/unittests/tests-gcoap/tests-gcoap.h:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * @brief   The entry point of this test suite.
+ * 
```
Please remove trailing whitespace.

kb2ma replied:
Fixed, apologies for the oversight.

miri64 commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -0,0 +1,181 @@
+                                                                           argv[4]);
+                }
+                printf("gcoap_cli: sending msg ID %u, %u bytes\n", coap_get_id(&pdu),
+                                                                   len);
```
`len` needs casting to `(unsigned)`

kb2ma replied:
Addressed and rebased.

miri64 commented on sys/net/gnrc/application_layer/coap/gcoap.c:
```diff
@@ -0,0 +1,597 @@
+    /* Content-Format */
+    if (pdu->content_type != COAP_FORMAT_NONE) {
+        bufpos += coap_put_option_ct(bufpos, last_optnum, pdu->content_type);
+        last_optnum = COAP_OPT_CONTENT_FORMAT;
```
last_optnum is assigned but never used according to `cppcheck`.

kb2ma replied:
Addressed and rebased.

kb2ma commented:
Now seeing an error in Murdock with dtls-echo:nrf52dk. I don't understand this, as nrf52dk is blacklisted in Makefile.

miri64 commented:
Nothing to do with your PR, so I would say ACK!

miri64 commented:
(See #6032)
